synapseutils/sync.py

```diff
@@ -585,88 +605,76 @@ def _order_items(
 
             graph[item.entity.path] = item_file_provenance
 
+        # Used to verify that the graph does not contain any cycles
         graph_sorted = utils.topolgical_sort(graph)
         results = {}
         for path, dependency_paths in graph_sorted:
             results.update({path: dependency_paths})
-        return results
 
-    async def upload(self, items: typing.Iterable[_SyncUploadItem]) -> None:
-        # Create dict of path -> File Entity
-        path_to_file_entity = {item.entity.path: item for item in items}
+        for file_path, dependent_file_paths in results.items():
+            dependent_tasks = []
+            for dependent_file in dependent_file_paths:
+                if resolved_file_checks.get(dependent_file, None):
+                    dependent_tasks.append(created_tasks_by_path.get(dependent_file))
+            if not created_tasks_by_path.get(file_path, None):
+                upload_item = items_by_path.get(file_path)
+                file_task = asyncio.create_task(
+                    self._upload_item_async(
+                        upload_item.entity,
+                        upload_item.used,
+                        upload_item.executed,
+                        upload_item.activity_name,
+                        upload_item.activity_description,
+                        dependent_tasks,
+                    )
+                )
+                created_tasks_by_path.update({file_path: file_task})
+
+        return created_tasks_by_path.values()
+
+    async def upload(self, items: Iterable[_SyncUploadItem]) -> None:
```

Comment:
Should this be `upload_async` so it can have a synchronous counterpart with the async->sync wrapper?

Reply:
I think I am going to make only the outer most entry point to the utility functions have an async and non-async version. So:

1. Sync to Synapse
2. Sync from Synapse
3. Generate manifest